A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+* The deserialized data can be attached to the metadata element, so that it can be deserialized once and accessed multiple times.
+* The deserialized data can be stored in the call data and accessed by the LB policy code when it intercepts the trailing metadata.
+
+In addition, we also discussed the possibility of having the LB policy get a separate callback for the load report data, 
+distinct from the callback it gets to intercept trailing metadata.  
+However, we think this is not a good idea, since it is likely to cause ordering problems
+(there is no guarantee of the order in which the callbacks would be seen by the LB policy) 
+and may incur additional synchronization overhead.
+
+#### The Server API 
+Using a server interceptor, it is convenient to attach the metrics as part of the
+trailing metadata of the server call and send back with the RPC response.
+
+Define a metric recorder that is unique per RPC. Server applications should obtain a reference and 
+use it to record metrics specifically for the current RPC call. Implementing this may leverage a 
+[context](https://grpc.github.io/grpc-java/javadoc/io/grpc/Context.html) or similar mechanism, 
+and a language should implement one if not already existed.
+The metric recoder looks like this:
+
+```Java
+public final class CallMetricRecorder {
+  // Records a request cost metric measurement for the call.
+  public CallMetricRecorder recordRequestCostMetric(String name, double value);
+
+  // Records a utilization metric measurement for the call.
+  public CallMetricRecorder recordUtilizationMetric(String name, double value);
+
+  //Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordCPUUtilizationMetric(double value);
+
+  // Records the CPU utilization metric measurement for the call. 
+  public CallMetricRecorder recordMemoryUtilizationMetric(double value);
+
+  // Returns the call metric recorder attached to the current call.
+  public static CallMetricRecorder getCurrent();
+}
+```
+
+All the methods should be thread safe.
+Recording one metrics multiple times overrides the previously provided values. 
+The methods can be called at any time during an RPC lifecycle,
+depending on the particular type of the cost or utilization measurement. For example, an RPC request size may
+be recorded at the beginning of the call, while a queue-size may be reported at any time during the call life-cycle.
+The data is dumped and translated to `OrcaLoadReport` at the end of each RPC by the interceptor, 
+and sent out as the trailing metadata.
+
+Note that reporting utilization metrics per-query is only useful in high QPS systems. 
+In such systems, the application that measures utilization data should consider sampling the data 
+to avoid performance degradation due to high rates of syscalls.
+
+
+### Out of Band Metrics Reporting
+For the out-of-band reporting, it requires a separate load reporting server running along with the 
```

Comment:
It does not require a separate *server* (i.e., listening on a different port); it requires a separate RPC *service* (i.e., a handler for requests for a particular RPC method).

I don't think this needs to say anything about "running along with the backend server for as long as the server is alive", since that seems irrelevant and confusing.